contracts/sdk/IpcContract.sol

```diff
@@ -1,151 +1,112 @@
-    /// before garbage collecting the in-flight message.
-    uint256 public timeoutBlocks;
+    // List of messages in-flight for which the contract hasn't received a receipt yet.
+    mapping(bytes32 => IpcEnvelope) public inflightMsgs;
 
-    constructor(address gatewayAddr_, uint256 timeoutBlocks_) {
+    constructor(address gatewayAddr_) {
         gatewayAddr = gatewayAddr_;
-        timeoutBlocks = timeoutBlocks_;
     }
 
-    error CallerIsNotGateway();
-    error IpcMsgCallFailed();
-    error UnsupportedMsgKind();
-    error UnrecognizedReceipt();
-
     /// @notice Entrypoint for IPC-enabled contracts. This function is always called by
     /// the gateway when a `Call` or `Receipt` cross-net messages is targeted to
     /// a specific address in the subnet.
-    function IpcEntrypoint(IpcEnvelope calldata crossMsg) external returns (bytes memory) {
-        // trigger the garbage collection of in-flight messages
-        // in case any timeout has triggered already.
-        // timeout == 0 means that no in-flight expiraton is set for receipts.
-        if (timeoutBlocks != 0) {
-            inFlightGC();
-        }
-
-        // only the gateway address is allowed to deliver cross-net messages.
-        if (msg.sender != gatewayAddr) {
-            revert CallerIsNotGateway();
-        }
-
+    function handleIpcMessage(IpcEnvelope calldata envelope) external payable onlyGateway returns (bytes memory) {
         // internal dispatch of the cross-net message to the right method.
-        if (crossMsg.kind == IpcMsgKind.Call) {
-            IpcMsg memory callMsg = abi.decode(crossMsg.message, (IpcMsg));
-
-            (bool success, bytes memory ret) = address(this).delegatecall(
-                abi.encodeWithSelector(callMsg.method, callMsg.params)
-            );
-            if (!success) {
-                revert IpcMsgCallFailed();
-            }
-
-            return ret;
-        } else if (crossMsg.kind == IpcMsgKind.Receipt) {
-            ReceiptMsg memory ackMsg = abi.decode(crossMsg.message, (ReceiptMsg));
-
-            // get the original message and chekc if the receipt was generated
-            // by the target contract, and that we actually sent that cross-net message.
-            // If this is not the case, revert, as someone may be trying to do something
-            // sketchy.
-            IpcEnvelope memory originalMsg = inFlightMsgs[ackMsg.id];
+        if (envelope.kind == IpcMsgKind.Call) {
+            CallMsg memory call = abi.decode(envelope.message, (CallMsg));
+            return _handleIpcCall(envelope, call);
+        } else if (envelope.kind == IpcMsgKind.Receipt) {
+            ResultMsg memory result = abi.decode(envelope.message, (ResultMsg));
+
+            // Recover the original message.
+            // If we were not tracking it, or if some details don't match, refuse to handle the receipt.
+            IpcEnvelope storage orig = inflightMsgs[result.id];
```

Comment:
There is no ay input validation for ID. ID is controlled by the sender.
Is it sufficient to check `keccak256(abi.encode(envelope.from)) != keccak256(abi.encode(orig.to))` ?

Reply:
What kind of input validation would make sense here? This method can only be invoked by the owner.

Reply:
I thought about removing arbitrary messages

Reply:
Could you elaborate further? I don't follow you.